katsdpcal/reduction.py

```diff
@@ -858,5 +915,45 @@ def flush_pipeline(ts, parameters, solution_stores):
     solution_stores : dict of :class:`~.CalSolutionStore`-like
         Solution stores for the capture block, indexed by solution type
     """
-    logger.info('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')
-    logger.info('Flushed pipeline')
+    # TODO: num_chunks should enter through paramters
+    #Group the frequency channels into this many sections
+    num_chunks = 16
+    # Extract bandpass gains from solution stores
+    b_solutions = solution_stores['B_POINTING'].get_range(start_time=0, end_time = time.time())
+    # Extract some some commonly used constants from the TS and parameters
+    # Middle time for each dump
+    mid_times = b_solutions.times
+    # TODO The following ideally needs katdal sensor framework for interpolation
+    target_str = ts.get_range('cbf_target', et = mid_times[0])[0][0]
+    target = katpoint.Target(target_str)
+    # Atmospheric conditions
+    soltime = np.mean(mid_times)
+    pres = ts.get_range('anc_air_pressure', et = soltime)[0][0]
+    temp = ts.get_range('anc_air_temperature', et = soltime)[0][0]
+    humi = ts.get_range('anc_air_relative_humidity', et = soltime)[0][0]
+    ants = parameters['antennas']
+    dump_period = ts['int_time']
+    channel_freqs = parameters['channel_freqs']
+    pols = parameters['pol_ordering']
+    
+    # Calculate offset (x,y) co-ordinates for each pointing
+    offsets = get_offsets(ts, parameters, target, mid_times, temp, pres, humi)
+    # Extract gains per pointing offset, per receptor and per frequency chunk.
+    data_points = pointing.get_offset_gains(b_solutions.values, offsets, ants, channel_freqs,
+                                             pols, num_chunks)
+    # Fit primary beams to the gains
+    beams = pointing.beam_fit(data_points, ants, num_chunks)
+    # Save fitted beams as CalSolution with shape (num_chunks, len(pols), len(ants), 5)
+    beam_sol = np.full((num_chunks, len(pols), len(ants), 5), np.nan, dtype=np.float32)
+    beam_sol_SNR = np.full((num_chunks, len(pols), len(ants), 5), np.nan, dtype=np.float32)
+    for a, ant in enumerate(ts.cal_antlist):
+        for c,beam in enumerate(beams[ant]):
+            beam_sol[c,:,a]=np.r_[beam.center, beam.width, beam.height]
+            beam_sol_SNR[c,:,a]=1/np.r_[beam.std_center, beam.std_width, beam.std_height]
```

Comment:
Just better Python styling:
```suggestion
            beam_sol_SNR[c, :, a] = 1 / np.r_[beam.std_center, beam.std_width, beam.std_height]
```